web/packages/teleterm/src/ui/services/deepLinks/deepLinksService.ts

```diff
@@ -108,6 +145,27 @@ export class DeepLinksService {
       return;
     }
 
+    const { isAtDesiredWorkspace, rootClusterUri } =
+      await this.loginAndSetActiveWorkspace(url);
+
+    if (!isAtDesiredWorkspace) {
+      return;
+    }
+
+    this.workspacesService
+      .getWorkspaceDocumentService(rootClusterUri)
+      .openConnectMyComputerDocument({ rootClusterUri });
+  }
+
+  /**
+   * loginAndSetActiveWorkspace will set the relevant cluster if it is in the app and, if not,
+   * it opens a login dialog with cluster address and username prefilled from the URL.
+   */
+  private async loginAndSetActiveWorkspace(url: DeepURL): Promise<{
+    isAtDesiredWorkspace: boolean;
+    rootClusterUri: RootClusterUri;
+    rootCluster?: Cluster;
+  }> {
```

Comment:
Let's do it like this, otherwise it'll be way too easy to forget about checking `isAtDesiredWorkspace` before consuming other fields. If we ever need to call `loginAndSetActiveWorkspace` and then ignore the result, we can reconsider this.

```typescript
  private async loginAndSetActiveWorkspace(url: DeepURL): Promise<
    | {
        isAtDesiredWorkspace: false;
      }
    | {
        isAtDesiredWorkspace: true;
        rootClusterUri: RootClusterUri;
        rootCluster?: Cluster;
      }
  > {
```

<details>
<summary>Full patch</summary>

```diff
diff --git a/web/packages/teleterm/src/ui/services/deepLinks/deepLinksService.ts b/web/packages/teleterm/src/ui/services/deepLinks/deepLinksService.ts
index c115747229..7f9659184d 100644
--- a/web/packages/teleterm/src/ui/services/deepLinks/deepLinksService.ts
+++ b/web/packages/teleterm/src/ui/services/deepLinks/deepLinksService.ts
@@ -108,13 +108,14 @@ export class DeepLinksService {
     id: string,
     token: string
   ): Promise<void> {
-    const { isAtDesiredWorkspace, rootCluster, rootClusterUri } =
-      await this.loginAndSetActiveWorkspace(url);
+    const result = await this.loginAndSetActiveWorkspace(url);
 
-    if (!isAtDesiredWorkspace) {
+    if (!result.isAtDesiredWorkspace) {
       return;
     }
 
+    const { rootCluster, rootClusterUri } = result;
+
     this.modalsService.openRegularDialog({
       kind: 'device-trust-authorize',
       rootClusterUri,
@@ -145,13 +146,14 @@ export class DeepLinksService {
       return;
     }
 
-    const { isAtDesiredWorkspace, rootClusterUri } =
-      await this.loginAndSetActiveWorkspace(url);
+    const result = await this.loginAndSetActiveWorkspace(url);
 
-    if (!isAtDesiredWorkspace) {
+    if (!result.isAtDesiredWorkspace) {
       return;
     }
 
+    const { rootClusterUri } = result;
+
     this.workspacesService
       .getWorkspaceDocumentService(rootClusterUri)
       .openConnectMyComputerDocument({ rootClusterUri });
@@ -161,11 +163,16 @@ export class DeepLinksService {
    * loginAndSetActiveWorkspace will set the relevant cluster if it is in the app and, if not,
    * it opens a login dialog with cluster address and username prefilled from the URL.
    */
-  private async loginAndSetActiveWorkspace(url: DeepURL): Promise<{
-    isAtDesiredWorkspace: boolean;
-    rootClusterUri: RootClusterUri;
-    rootCluster?: Cluster;
-  }> {
+  private async loginAndSetActiveWorkspace(url: DeepURL): Promise<
+    | {
+        isAtDesiredWorkspace: false;
+      }
+    | {
+        isAtDesiredWorkspace: true;
+        rootClusterUri: RootClusterUri;
+        rootCluster?: Cluster;
+      }
+  > {
     const rootClusterId = url.hostname;
     const clusterAddress = url.host;
     const prefill = {
@@ -191,7 +198,6 @@ export class DeepLinksService {
       if (canceled) {
         return {
           isAtDesiredWorkspace: false,
-          rootClusterUri,
         };
       }
     }

```
</details>